### ISSUE TITLE
RFE: Bootstrap Checkpointing - Modify manifest behavior slightly for self hosting. 

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#49236]
**FEATURE REQUEST**:

As of today current self hosting model is as follows: 
1. Load the Static Pod specification from disk (from /etc/kubernetes/manifests)
2. Extract the PodSpec from that Static Pod specification
3. Mutate the PodSpec to be compatible with self-hosting (add the right labels, taints, etc. so it can schedule correctly)
4. Build a new DaemonSet object for the self-hosted component in question. Use the above mentioned PodSpec
5. Create the DaemonSet resource. Wait until the Pods are running.
6. Remove the Static Pod manifest file. The kubelet will stop the original Static Pod-hosted component that was running.
7. The self-hosted containers should now step up and take over.
...

A problem arises when we have a power failure condition.  Because we no longer have a manifest of what was running we can no longer restart the control-plane components. 

A very simple workaround would be the following: 
1. Add a --checkpoint-path flag with package defaults to /var/lib/kubernetes/checkpoints such that the kubelet can safely write to this location and follow standard conventions. 
2. On startup, have the kubelet read in the manifest-path as usual. 
3. Also read in the --checkpoint-path if specified
4. Prior to pod startup:
   A. Check the pod for a well defined annotation or label.
   A. delete old checkpoints (if there is one) 
   B. write named_YYYYMMDDHHMM manifest to the checkpoint location.

This way the kubelet keeps an active record of it's running pods.  There are other complexities around various other checkpointing ideas, and we've (@luxas and myself) decided that this is the bare minimum change that is simple and will enable us to move forwards. 

TODO: We will likely need a minor iteration for configmaps as well. 

/kind feature
/cc @kubernetes/sig-node-feature-requests @kubernetes/sig-cluster-lifecycle-feature-requests 
@roberthbailey @dchen1107 @luxas @jbeda

**Comment by luxas** [comment]
Can we please keep `--pod-manifest-path` and the checkpointing location separate @timothysc?
I think that would make sense, right now I don't have a clear idea how those interact with each other...

**Comment by luxas** [comment]
Because this is an alpha feature and `--pod-manifest-path` is very well-known to be `/etc/kubernetes/manifests` (I don't want to change that) and read-only for the kubelet.

Changing this in one go seems like a drastic change IMO. I think it's much better to separate from where the kubelet reads manifests that are known to be admin-only maintained and read-only for the kubelet and have a separate dir that the kubelet owns completely. We don't know what such a change would lead to for upgradability either.

The logic under the hood could be the same, the kubelet is just watching two directories in "read-only" mode, but can write to the checkpointing dir.

**Comment by timothysc** [comment]
I dunno about "drastic" but I can buy your argument.   ;-)

**Comment by dchen1107** [comment]
@yujuhong

**Comment by yujuhong** [comment]
I synced up with @timothysc. We agree that for 1.8, the scope of checkpointing should be limited to pods with specific annotations, and they will be persisted to disk.

Here are my other thoughts. I think the plan sounds reasonable, but it's designed for a very specific use case. I'd like to see a more generic mechanism that may be extended to checkpoint all pods and more (e.g., other resources, kubelet's internal states, etc) in the future. More specifically, this proposal 
  * Use pod manifests file. This cannot be used to store other non-k8s-API objects, and has no checksum to detect corruption.
  * Checkpoint only "Running" pods. This essentially checkpoints kubelet's internal state since it requires the pods to be running.

I propose that we 
  * Checkpoint pods "received" from the the apiserver (i.e., no kubelet's internal states are checkpointed)
  * At kubelet startup, load the checkpoint into kubelet's in-memory cache.
  * Once kubelet connects to the apiserver, it will update the in-memory cache AND the checkpoint on the disk.

The downside of this approach is that we cannot handle secrets/configmaps (yet) without additional work. Since those objects are already out of the scope for 1.8 (and not required by kubeadm), I think it's reasonable. Also note that the initial scope only includes pods from the apiserver (and not other sources such as http), but it should be able to extend this to other sources.

**Comment by aaronlevy** [comment]
I would really hope that we can keep checkpointing secrets/configMaps as in-scope for the initial checkpointing implementation.

Without these, it drastically reduces the utility of checkpointing to the point that self-hosted installations widely used in production ([Bootkube](https://github.com/kubernetes-incubator/bootkube) / [Tectonic](https://github.com/coreos/tectonic-installer)) - can not start evaluating or making use of this feature.

This feature only becomes useful if you assume assets that a pod needs will exist on the local node. This assumption, in my opinion, negates much of the value of self-hosting / checkpointing.

The reason that we use secrets and configmaps in our self-hosted control-plane is to decouple the application from the local node state. If our applications are coupled to the local node state, then we still have a hard requirement on external tooling to provide that state (VS just kubernetes objects).

We additionally want to ensure we're future-proofing ourselves for use of ComponentConfigs as a means for dynamic configuration of core kubernetes services (and not relying on static locally stored state).

**Comment by luxas** [comment]
@aaronlevy As discussed in the more implementation-focused meetings, we have plans to checkpoint ConfigMaps, but it's a stretch goal for v1.8. If it's really important to CoreOS, can you have someone help out with that?

Re checkpointing Secrets we concluded that we need to be _really_ careful here. What we're saying with Secrets is that they will be stored only in one place -- etcd. When landing on a node, they will be in RAM (tmpfs). If we suddenly start checkpointing Secrets by writing those to disk, we're effectively breaking a v1 Secret promise.

**Comment by kfox1111** [comment]
I think its ok to put an asterisk in the promise and state that it doesn't apply when checkpoints are used. It will not be used on very many pods and it will be invaluable to those that need it. If it isn't supported, people will just cram secrets into configmaps, which i think is much worse. :/

**Comment by aaronlevy** [comment]
Checkpointing secrets essentially becomes optional. If you add the checkpoint annotation to a pod that contains a secret - it is checkpointed. If you don't want secrets to be checkpointed - don't use a secret on a pod you are checkpointing.

I don't get the difference between checkpointing the secret to disk -- or just storing that secret on disk anyway. It's going to hit disk in both cases. If we wanted to we could do something like encrypt the at-rest secret using the kubelet's private key -- if that was exposed, you could request the secret from the apiserver directly anyway. But I still don't get the "it can't touch disk" argument if it is going to end up there anyway - and it would be optional.

**Comment by kfox1111** [comment]
@aaron12134 I think the concern is secrets only ever hit /dev/shm now, not disk ever.

I do think its a reasonable to allow it on disk when the pod is explicitly marked by the user as checkpointed though. Its critical to the functioning of the checkpoint that the pod can come up without the control plane being ready.

**Comment by aaronlevy** [comment]
@kfox1111 I mean that in this proposal, for example, the apiserver would need TLS assets. Those assets could be in a secret object that is checkpointed to disk. Or, if that's not supported -- those assets still have to exist on disk anyway, and be mounted into the apiserver pod. In no case are those assets not on disk. I might be missing something - but I don't see the fundamental difference.

**Comment by timothysc** [comment]
@aaronlevy I'm totally ok with baby stepping this across releases and locking it behind a feature gate.  It's a stretch goal to hit configmaps and a non-goal to hit secrets (at this time).  We can always chat about this more on Friday as well.  There was a consensus against storing certs as secrets vs. host volume mounts on the masters.

**Comment by ericchiang** [comment]
cc @rphillips

**Comment by aaronlevy** [comment]
Sure - happy to discuss further on Friday. My understanding was that kubeadm folks had decided not to use secrets, and I have no objections to that (it should be optional).

I also don't disagree that development could come in phases, but my position would be that support for checkpointing configMaps and secrets should be part of the proposal (and designed into the architecture of the feature).

In my mind, one of the main drivers of self-hosting components is that we are removing the need for on-host configuration external to k8s. If we're assuming that we have to configure assets on the host via an external tool - we're not moving the needle much from managing static manifests directly on the host anyway.

To be completely fair, this is also coming from a selfish position in that I want to retire the pod-checkpointer and move bootkube to using upstream kubelet checkpointing - and I would be infinitely saddened if kubelet checkpointing finally came about, and we couldn't use it.

**Comment by aaronlevy** [comment]
tl;dr: Would it be possible to have some representatives from sig-node and sig-auth join the working-group meeting next week to discuss inclusion of configmaps / secrets as part of the checkpointing proposal? Hoping to get some consensus on what could be in scope, and discuss associated risks/concerns.

I didn't want to ping entire sig lists, so @yujuhong and @ericchiang -- any particular people that might be good to ping from sig-node/sig-auth (or whole sigs)? The meeting time next week is Wednesday 16:00UTC / 9am PT.

Then as a general follow up from today's meeting:

- [Meeting notes are here](https://docs.google.com/document/d/1deJYPIF4LmhGjDVaqrswErIrV7mtwJgovtLnPCDxP7U) (end of section discusses checkpointing).
- [Current proposal](https://docs.google.com/document/d/1hhrCa_nv0Sg4O_zJYOnelE8a5ClieyewEsQM6c7-5-o) 

My overal perspective is that a goal of self-hosting is that applications ideally do not require any knowledge of the underlying host (or require external tools to configure particular hosts in a "special" way). A master node, for example, would be modeled through labels and taints, not on-host configuration steps. And control-plane applications have all of their dependencies distributed as Kubernetes objects.

To this end, my concern is that if we completely leave out the discussion of checkpointing 
ConfigMaps and Secrets as part of this proposal, we are (in my mind) putting the utility of this feature at risk. I worry that if we don't get buy-in on supporting these as part of pod checkpointing, it's going to be much more difficult to go back and essentially re-scope the feature later.

**Comment by yujuhong** [comment]
> To be completely fair, this is also coming from a selfish position in that I want to retire the pod-checkpointer and move bootkube to using upstream kubelet checkpointing - and I would be infinitely saddened if kubelet checkpointing finally came about, and we couldn't use it.

+1 for retiring the pod-cehckpointer eventually. I looked at how it worked briefly last week, and I saw several issues with it (e.g., relying on a debug /runningpods endpoint).
That said, I don't think we need to rush to support configmap and secret checkpointing in 1.8. The code freeze is about a month away, and we've yet to agree on a detailed proposal. I think we should be realistic and limit the scope if we want to achieve anything for 1.8 at all. Since bootkube has been widely used in production for a while, I'd guess it's not urgent to migrate bootkube right now. 

If we are not targeting this for 1.8, I agree that we can expand the scope and do more. I'd also like to point out that checkpointing in kubelet can solve more issues than self-hosting master components, and we (or at least some sig-node members) intend to leverage the mechanism in the future. We'd be rather disappointed to see a special solution developed just for this use case. 


> I didn't want to ping entire sig lists, so @yujuhong and @ericchiang -- any particular people that might be good to ping from sig-node/sig-auth (or whole sigs)? The meeting time next week is Wednesday 16:00UTC / 9am PT.

I can attend the meeting. I'll ask to see if anyone else from sig-node is interested.

**Comment by xiang90** [comment]
@yujuhong 

> That said, I don't think we need to rush to support configmap and secret checkpointing in 1.8.

What are the potential challenges of configmap checkpointing? Is it possible for us to include the design of configmap in the initial proposal?

**Comment by ericchiang** [comment]
> I didn't want to ping entire sig lists, so @yujuhong and @ericchiang -- any particular people that might be good to ping from sig-node/sig-auth (or whole sigs)? The meeting time next week is Wednesday 16:00UTC / 9am PT.

I'll join as well.

cc @kubernetes/sig-auth-proposals

**Comment by yujuhong** [comment]
Can anyone post the meeting information somewhere?

**Comment by luxas** [comment]
@yujuhong Here is the general SIG Cluster Lifecycle Meeting Notes: https://docs.google.com/document/d/1deJYPIF4LmhGjDVaqrswErIrV7mtwJgovtLnPCDxP7U/edit#
The more specialized meeting we're going to have tomorrow has its meeting notes here: https://docs.google.com/document/d/16CEsBSSGm3sMpvB_cFnKnqqi1OxhIcyX3lVwBpIyMHc/edit (they are linked to each other though)

**Comment by roberthbailey** [comment]
Moving to the 1.9 milestone as this isn't going to land before code freeze.

**Comment by luxas** [comment]
@timothysc Marked this approved for v1.9.
Do we have any recent progress?

**Comment by timothysc** [comment]
Conversed with @yujuhong today and I'll get moving on it tomorrow.

**Comment by dims** [comment]
@timothysc @yujuhong - punt this to 1.10?

**Comment by kfox1111** [comment]
Its close to merging isnt it?
